// @synapseprotocol/sdk/dist/hooks/bridge.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { Bridge } from "../bridge/bridge.js";
import { useSignerFromEthereum } from "./signer.js";
import { logError, parseBigNumberish } from "./helpers.js";
import { useApproveStatus, useCheckAllowance } from "./tokens.js";
import { useEffect, useState } from "react";
function useCalculateBridgeSwapOutput(args) {
    var ethereum = args.ethereum, chainId = args.chainId, rest = __rest(args, ["ethereum", "chainId"]);
    var _a = useState(null), result = _a[0], setResult = _a[1];
    function fn() {
        var fnArgs = __assign(__assign({}, rest), { amountFrom: parseBigNumberish(rest.amountFrom, rest.tokenFrom, chainId) });
        var synapseBridge = new Bridge.SynapseBridge({ network: chainId });
        synapseBridge.estimateBridgeTokenOutput(fnArgs)
            .then(setResult)
            .catch(logError);
    }
    return [fn, result];
}
function useExecuteBridgeSwap(args) {
    var ethereum = args.ethereum, chainId = args.chainId, rest = __rest(args, ["ethereum", "chainId"]);
    var getSigner = useSignerFromEthereum()[0];
    var _a = useState(null), tx = _a[0], setTx = _a[1];
    function fn() {
        var fnArgs = __assign(__assign({}, rest), { amountFrom: parseBigNumberish(rest.amountFrom, rest.tokenFrom, chainId), amountTo: parseBigNumberish(rest.amountTo, rest.tokenTo, rest.chainIdTo) });
        var synapseBridge = new Bridge.SynapseBridge({ network: chainId });
        synapseBridge.executeBridgeTokenTransaction(fnArgs, getSigner(ethereum))
            .then(setTx)
            .catch(logError);
    }
    return [fn, tx];
}
function useApproveBridgeSwap(args) {
    var ethereum = args.ethereum, chainId = args.chainId, rest = __rest(args, ["ethereum", "chainId"]);
    var getSigner = useSignerFromEthereum()[0];
    var _a = useApproveStatus(ethereum, chainId), queryApproveStatus = _a[0], approvalStatus = _a[1];
    var _b = useState(null), approveTx = _b[0], setApproveTx = _b[1], _c = useState(null), approveData = _c[0], setApproveData = _c[1];
    var amt = rest.amount
        ? parseBigNumberish(rest.amount, rest.token, chainId)
        : undefined;
    var fnArgs = __assign(__assign({}, rest), { amount: amt });
    function fn() {
        var synapseBridge = new Bridge.SynapseBridge({ network: chainId });
        var spender = synapseBridge.buildERC20ApproveArgs(fnArgs)[0].spender;
        synapseBridge.executeApproveTransaction(fnArgs, getSigner(ethereum))
            .then(function (res) {
            setApproveTx(res);
            setApproveData(__assign(__assign({}, fnArgs), { spender: spender }));
        })
            .catch(logError);
    }
    useEffect(function () {
        if (approvalStatus !== null) {
            return;
        }
        if ((approveTx && approveData)) {
            var token = approveData.token, spender = approveData.spender, amount = approveData.amount;
            queryApproveStatus({
                token: token,
                spender: spender,
                amount: amount,
                approveTx: approveTx
            });
        }
    }, [approveTx, approveData, approvalStatus]);
    return [fn, approveTx, approvalStatus];
}
function useBridgeAllowance(args) {
    var ethereum = args.ethereum, chainId = args.chainId, token = args.token;
    var _a = useCheckAllowance(ethereum, chainId), checkAllowance = _a[0], allowance = _a[1];
    useEffect(function () {
        var synapseBridge = new Bridge.SynapseBridge({ network: chainId });
        var spender = synapseBridge.buildERC20ApproveArgs({ token: token })[0].spender;
        checkAllowance({ token: token, spender: spender });
    }, [chainId, token]);
    return [allowance];
}
function useNeedsBridgeSwapApproval(args) {
    var chainId = args.chainId, token = args.token, amount = args.amount;
    var amt = parseBigNumberish(amount, token, chainId);
    var allowance = useBridgeAllowance(args)[0];
    var _a = useState(null), needsApprove = _a[0], setNeedsApprove = _a[1];
    useEffect(function () {
        if (allowance) {
            setNeedsApprove(allowance.lt(amt));
        }
    }, [allowance, chainId, token, amount]);
    return [needsApprove, allowance];
}
function useBridgeSwapApproval(args) {
    var _a = useNeedsBridgeSwapApproval(args), needsApprove = _a[0], allowance = _a[1];
    var _b = useApproveBridgeSwap(__assign(__assign({}, args), { amount: undefined })), execApprove = _b[0], approveTx = _b[1], approveStatus = _b[2];
    return {
        needsApprove: needsApprove,
        allowance: allowance,
        execApprove: execApprove,
        approveTx: approveTx,
        approveStatus: approveStatus
    };
}
export { useApproveBridgeSwap, useExecuteBridgeSwap, useCalculateBridgeSwapOutput, useBridgeAllowance, useNeedsBridgeSwapApproval, useBridgeSwapApproval };
