// @synapseprotocol/sdk/dist/bridge/gasutils.js
import { ChainId } from "../common/chainid.js";
import { populateGasOptions } from "../common/gasoptions.js";
import { parseUnits } from "@ethersproject/units";
import { BigNumber } from "@ethersproject/bignumber";
export var GasUtils;
(function (GasUtils) {
    var _a;
    var makeGwei = function (n) { return parseUnits(n, "gwei"); };
    var CHAIN_GAS_PARAMS = (_a = {},
        _a[ChainId.ETH] = {
            maxFeePerGas: makeGwei("80"),
            maxPriorityFee: makeGwei("1.5"),
            bridgeGasLimit: BigNumber.from(301000),
            approveGasLimit: BigNumber.from(75000),
        },
        _a[ChainId.OPTIMISM] = {
            bridgeGasLimit: BigNumber.from(250000),
            approveGasLimit: BigNumber.from(90000),
        },
        _a[ChainId.BSC] = {
            gasPrice: makeGwei("6"),
            bridgeGasLimit: BigNumber.from(260000),
            approveGasLimit: BigNumber.from(75000)
        },
        _a[ChainId.POLYGON] = {
            maxFeePerGas: makeGwei("32.01"),
            maxPriorityFee: makeGwei("32"),
            bridgeGasLimit: BigNumber.from(1000000),
            approveGasLimit: BigNumber.from(86000)
        },
        _a[ChainId.BOBA] = {
            gasPrice: makeGwei("10"),
            approveGasLimit: BigNumber.from(60000),
        },
        _a[ChainId.ARBITRUM] = {
            gasPrice: makeGwei("1.5"),
            bridgeGasLimit: BigNumber.from(1500000),
        },
        _a[ChainId.AVALANCHE] = {
            maxFeePerGas: makeGwei("210"),
            maxPriorityFee: makeGwei("3"),
            bridgeGasLimit: BigNumber.from(700000),
            approveGasLimit: BigNumber.from(75000),
        },
        _a[ChainId.AURORA] = {
            gasPrice: makeGwei('1'),
        },
        _a[ChainId.HARMONY] = {
            gasPrice: makeGwei("45"),
            bridgeGasLimit: BigNumber.from(250000),
            approveGasLimit: BigNumber.from(75000)
        },
        _a);
    function approveLimit(chainId) {
        var _a;
        if (chainId in CHAIN_GAS_PARAMS) {
            var gasParams = CHAIN_GAS_PARAMS[chainId];
            return (_a = gasParams.approveGasLimit) !== null && _a !== void 0 ? _a : null;
        }
        return null;
    }
    GasUtils.approveLimit = approveLimit;
    function bridgeGasLimit(chainId) {
        var _a;
        if (chainId in CHAIN_GAS_PARAMS) {
            var gasParams = CHAIN_GAS_PARAMS[chainId];
            return (_a = gasParams.bridgeGasLimit) !== null && _a !== void 0 ? _a : null;
        }
        return null;
    }
    GasUtils.bridgeGasLimit = bridgeGasLimit;
    GasUtils.makeGasParams = function (chainId) { var _a; return (_a = CHAIN_GAS_PARAMS[chainId]) !== null && _a !== void 0 ? _a : {}; };
    GasUtils.populateGasParams = function (chainId, txn, gasLimitKind) {
        return Promise.resolve(txn)
            .then(function (tx) {
            var _a = GasUtils.makeGasParams(chainId), maxFeePerGas = _a.maxFeePerGas, maxPriorityFee = _a.maxPriorityFee, gasPrice = _a.gasPrice, approveGasLimit = _a.approveGasLimit, bridgeGasLimit = _a.bridgeGasLimit;
            tx.chainId = chainId;
            var gasOpts = {};
            if (gasPrice) {
                gasOpts.gasPrice = gasPrice;
            }
            else if (maxFeePerGas) {
                gasOpts.maxFeePerGas = maxFeePerGas;
                if (maxPriorityFee) {
                    gasOpts.maxPriorityFeePerGas = maxPriorityFee;
                }
            }
            switch (gasLimitKind) {
                case "bridge":
                    if (bridgeGasLimit)
                        gasOpts.gasLimit = bridgeGasLimit;
                    break;
                case "approve":
                    if (approveGasLimit)
                        gasOpts.gasLimit = approveGasLimit;
                    break;
            }
            return populateGasOptions(tx, gasOpts, chainId);
        });
    };
})(GasUtils || (GasUtils = {}));
