// @synapseprotocol/sdk/dist/common/utils.js
var _a;
import { ChainId } from "./chainid.js";
import { SynapseContracts } from "./synapse_contracts.js";
import { StaticCallResult } from "./types.js";
import { BigNumber } from "@ethersproject/bignumber";
export function rejectPromise(e) { return Promise.reject(e instanceof Error ? e : new Error(e)); }
export function executePopulatedTransaction(populatedTxn, signer) {
    return Promise.resolve(populatedTxn)
        .then(function (txn) { return signer.sendTransaction(txn); })
        .catch(rejectPromise);
}
export function staticCallPopulatedTransaction(populatedTxn, signer) {
    return Promise.resolve(populatedTxn)
        .then(function (txn) {
        return signer.call(txn)
            .then(function () { return StaticCallResult.Success; })
            .catch(function (err) { return StaticCallResult.Failure; });
    });
}
export function pow10(exp) { return BigNumber.from(10).pow(exp); }
/**
 * "Fixes" a value into units of Wei; should be used when tokens
 * have a decimals value which isn't 18
 * (such as USDC/USDT on chains which aren't BSC) and you need to do
 * calculations using proper units of Wei instead of, for example in the case of
 * USDC/USDT, Szabo (10^-6)
 * @param amt
 * @param decimals
 */
export function fixWeiValue(amt, decimals) {
    var multiplier = pow10(18).div(pow10(decimals));
    return amt.mul(multiplier);
}
export function contractAddressFor(chainId, key) {
    var address;
    var contractsForChain = contractsForChainId(chainId);
    switch (key) {
        case "bridgeAddress":
            address = contractsForChain.bridgeAddress;
            break;
        case "bridgeZapAddress":
            address = contractsForChain.bridgeZapAddress;
            break;
    }
    return address;
}
var CHAINID_CONTRACTS_MAP = (_a = {},
    _a[ChainId.ETH] = SynapseContracts.Ethereum,
    _a[ChainId.OPTIMISM] = SynapseContracts.Optimism,
    _a[ChainId.CRONOS] = SynapseContracts.Cronos,
    _a[ChainId.BSC] = SynapseContracts.BSC,
    _a[ChainId.POLYGON] = SynapseContracts.Polygon,
    _a[ChainId.FANTOM] = SynapseContracts.Fantom,
    _a[ChainId.BOBA] = SynapseContracts.Boba,
    _a[ChainId.METIS] = SynapseContracts.Metis,
    _a[ChainId.MOONBEAM] = SynapseContracts.Moonbeam,
    _a[ChainId.MOONRIVER] = SynapseContracts.Moonriver,
    _a[ChainId.ARBITRUM] = SynapseContracts.Arbitrum,
    _a[ChainId.AVALANCHE] = SynapseContracts.Avalanche,
    _a[ChainId.DFK] = SynapseContracts.DFK,
    _a[ChainId.AURORA] = SynapseContracts.Aurora,
    _a[ChainId.HARMONY] = SynapseContracts.Harmony,
    _a);
export var contractsForChainId = function (chainId) { var _a; return (_a = CHAINID_CONTRACTS_MAP[chainId]) !== null && _a !== void 0 ? _a : null; };
