// @synapseprotocol/sdk/dist/token.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { BigNumber } from "@ethersproject/bignumber";
import { formatEther, parseUnits } from "@ethersproject/units";
export function instanceOfToken(object) {
    return object instanceof BaseToken || object instanceof WrapperToken;
}
/**
 * Token represents an ERC20 token on Ethereum-based blockchains.
 */
var BaseToken = /** @class */ (function () {
    /**
     * Creates a new Token object with the defined arguments.
     * @param {Object} args Information about this token, including name, symbol, decimals, and
     * contract addresses.
     * @param {string} args.name Name of the token (example, "USD Circle")
     * @param {string} args.symbol Symbol of the token (example, "USDC")
     * @param {number|Object} args.decimals Either a single value, representing the token's ERC20 decimals value on all chains, or
     * a map in the format of { chain id => decimals for chain }.
     * If the latter is passed, values for ALL known chains must be provided.
     * @param {Object} args.addresses Mapping in the format of { chain id => address of token on chain },
     * providing the address of this token on different chains.
     * @param {SwapType} args.swapType Swap type of this token
     */
    function BaseToken(args) {
        var _a, _b, _c;
        this.addresses = {};
        this.wrapperAddresses = {};
        this._decimals = {};
        this.name = args.name;
        this.symbol = args.symbol;
        this.addresses = args.addresses;
        this.swapType = args.swapType;
        this.wrapperAddresses = (_a = args.wrapperAddresses) !== null && _a !== void 0 ? _a : {};
        if (typeof args.decimals === "number") {
            for (var _i = 0, _d = Object.entries(this.addresses); _i < _d.length; _i++) {
                var k = _d[_i][0];
                this._decimals[k] = args.decimals;
            }
        }
        else {
            this._decimals = args.decimals;
        }
        this.isETH = (_b = args.isETH) !== null && _b !== void 0 ? _b : false;
        this.isGasToken = (_c = args.isGasToken) !== null && _c !== void 0 ? _c : false;
        this.id = Symbol(this.symbol);
    }
    Object.defineProperty(BaseToken.prototype, "isWrapperToken", {
        get: function () {
            return false;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Returns the address of this token on a given network, or null if
     * the token does not exist on the passed network.
     * @param {number} chainId Chain ID
     * @return {string|null} Token's contract address for the queried network, or null
     */
    BaseToken.prototype.address = function (chainId) {
        return this.addresses[chainId] || null;
    };
    BaseToken.prototype.wrapperAddress = function (chainId) {
        return this.wrapperAddresses[chainId] || null;
    };
    BaseToken.prototype.decimals = function (chainId) {
        return this._decimals[chainId] || null;
    };
    BaseToken.prototype.isEqual = function (other) {
        return this.id === other.id;
    };
    BaseToken.prototype.weiToEther = function (amt, chainId) {
        var decimals = this.decimals(chainId) || 18, multiplier = BigNumber.from(10).pow(18 - decimals);
        return BigNumber.from(amt).mul(multiplier);
    };
    BaseToken.prototype.weiToEtherString = function (amt, chainId) {
        var decimals = this.decimals(chainId) || 18, multiplier = BigNumber.from(10).pow(18 - decimals);
        return formatEther(BigNumber.from(amt).mul(multiplier));
    };
    BaseToken.prototype.etherToWei = function (amt, chainId) {
        var _a;
        var etherStr = amt instanceof BigNumber
            ? amt.toString()
            : BigNumber.from(amt).toString();
        return parseUnits(etherStr, (_a = this.decimals(chainId)) !== null && _a !== void 0 ? _a : 18);
    };
    BaseToken.prototype.valueToWei = function (ether, chainId) {
        return this.etherToWei(ether, chainId);
    };
    BaseToken.prototype.canSwap = function (other) {
        return this.swapType === other.swapType;
    };
    return BaseToken;
}());
export { BaseToken };
var WrapperToken = /** @class */ (function (_super) {
    __extends(WrapperToken, _super);
    function WrapperToken(args) {
        var _this = _super.call(this, args) || this;
        _this.underlyingToken = args.underlyingToken;
        return _this;
    }
    Object.defineProperty(WrapperToken.prototype, "isWrapperToken", {
        get: function () {
            return true;
        },
        enumerable: false,
        configurable: true
    });
    return WrapperToken;
}(BaseToken));
export { WrapperToken };
