// @synapseprotocol/sdk/dist/tokens.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { BaseToken, WrapperToken } from "./token.js";
import { ChainId } from "./common/chainid.js";
import { ERC20 } from "./bridge/erc20.js";
import { SwapType } from "./internal/swaptype.js";
import { tokenSwitch } from "./internal/utils.js";
export var Tokens;
(function (Tokens) {
    // Stablecoins
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21;
    /**
     * The DAI stablecoin, available on Arbitrum, Avalanche, Binance Smart Chain, Ethereum, and Polygon.
     */
    Tokens.DAI = new BaseToken({
        name: "Dai",
        symbol: "DAI",
        decimals: 18,
        addresses: (_a = {},
            _a[ChainId.ETH] = "0x6b175474e89094c44da98b954eedeac495271d0f",
            _a[ChainId.CRONOS] = "0xf2001b145b43032aaf5ee2884e456ccd805f677d",
            _a[ChainId.BSC] = "0x1af3f329e8be154074d8769d1ffa4ee058b1dbc3",
            _a[ChainId.POLYGON] = "0x8f3cf7ad23cd3cadbd9735aff958023239c6a063",
            _a[ChainId.BOBA] = "0xf74195Bb8a5cf652411867c5C2C5b8C2a402be35",
            _a[ChainId.ARBITRUM] = "0xDA10009cBd5D07dd0CeCc66161FC93D7c9000da1",
            _a[ChainId.AVALANCHE] = "0xd586E7F844cEa2F87f50152665BCbc2C279D8d70",
            _a[ChainId.AURORA] = "0xe3520349F477A5F6EB06107066048508498A291b",
            _a[ChainId.HARMONY] = "0xef977d2f931c1978db5f6747666fa1eacb0d0339",
            _a),
        swapType: SwapType.USD
    });
    Tokens.BUSD = new BaseToken({
        name: "Binance USD",
        symbol: "BUSD",
        decimals: 18,
        addresses: (_b = {},
            _b[ChainId.BSC] = "0xe9e7cea3dedca5984780bafc599bd69add087d56",
            _b),
        swapType: SwapType.USD
    });
    Tokens.USDC = new BaseToken({
        name: "USD Circle",
        symbol: "USDC",
        decimals: (_c = {},
            _c[ChainId.ETH] = 6,
            _c[ChainId.OPTIMISM] = 6,
            _c[ChainId.CRONOS] = 6,
            _c[ChainId.BSC] = 18,
            _c[ChainId.POLYGON] = 6,
            _c[ChainId.FANTOM] = 6,
            _c[ChainId.BOBA] = 6,
            _c[ChainId.METIS] = 6,
            _c[ChainId.AVALANCHE] = 6,
            _c[ChainId.ARBITRUM] = 6,
            _c[ChainId.AURORA] = 6,
            _c[ChainId.HARMONY] = 6,
            _c),
        addresses: (_d = {},
            _d[ChainId.ETH] = "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
            _d[ChainId.OPTIMISM] = "0x7f5c764cbc14f9669b88837ca1490cca17c31607",
            _d[ChainId.CRONOS] = "0xc21223249ca28397b4b6541dffaecc539bff0c59",
            _d[ChainId.BSC] = "0x8ac76a51cc950d9822d68b83fe1ad97b32cd580d",
            _d[ChainId.POLYGON] = "0x2791bca1f2de4661ed88a30c99a7a9449aa84174",
            _d[ChainId.FANTOM] = "0x04068da6c83afcfa0e13ba15a6696662335d5b75",
            _d[ChainId.BOBA] = "0x66a2A913e447d6b4BF33EFbec43aAeF87890FBbc",
            _d[ChainId.METIS] = "0xEA32A96608495e54156Ae48931A7c20f0dcc1a21",
            _d[ChainId.ARBITRUM] = "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
            _d[ChainId.AVALANCHE] = "0xA7D7079b0FEaD91F3e65f86E8915Cb59c1a4C664",
            _d[ChainId.AURORA] = "0xB12BFcA5A55806AaF64E99521918A4bf0fC40802",
            _d[ChainId.HARMONY] = "0x985458e523db3d53125813ed68c274899e9dfab4",
            _d),
        swapType: SwapType.USD
    });
    Tokens.USDT = new BaseToken({
        name: "USD Tether",
        symbol: "USDT",
        decimals: (_e = {},
            _e[ChainId.ETH] = 6,
            _e[ChainId.CRONOS] = 6,
            _e[ChainId.BSC] = 18,
            _e[ChainId.POLYGON] = 6,
            _e[ChainId.FANTOM] = 6,
            _e[ChainId.BOBA] = 6,
            _e[ChainId.ARBITRUM] = 6,
            _e[ChainId.AVALANCHE] = 6,
            _e[ChainId.AURORA] = 6,
            _e[ChainId.HARMONY] = 6,
            _e),
        addresses: (_f = {},
            _f[ChainId.ETH] = "0xdac17f958d2ee523a2206206994597c13d831ec7",
            _f[ChainId.CRONOS] = "0x66e428c3f67a68878562e79a0234c1f83c208770",
            _f[ChainId.BSC] = "0x55d398326f99059ff775485246999027b3197955",
            _f[ChainId.POLYGON] = "0xc2132d05d31c914a87c6611c10748aeb04b58e8f",
            _f[ChainId.FANTOM] = "0x049d68029688eabf473097a2fc38ef61633a3c7a",
            _f[ChainId.BOBA] = "0x5DE1677344D3Cb0D7D465c10b72A8f60699C062d",
            _f[ChainId.ARBITRUM] = "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
            _f[ChainId.AVALANCHE] = "0xc7198437980c041c805a1edcba50c1ce5db95118",
            _f[ChainId.AURORA] = "0x4988a896b1227218e4A686fdE5EabdcAbd91571f",
            _f[ChainId.HARMONY] = "0x3c2b8be99c50593081eaa2a724f0b8285f5aba8f",
            _f),
        swapType: SwapType.USD
    });
    Tokens.UST = new BaseToken({
        symbol: "UST",
        name: "TerraUSD",
        decimals: 18,
        addresses: (_g = {},
            _g[ChainId.ETH] = "0x0261018Aa50E28133C1aE7a29ebdf9Bd21b878Cb",
            _g[ChainId.OPTIMISM] = "0xFB21B70922B9f6e3C6274BcD6CB1aa8A0fe20B80",
            _g[ChainId.CRONOS] = "0x7Bb5c7e3bF0B2a28fA26359667110bB974fF9359",
            _g[ChainId.BSC] = "0xb7A6c5f0cc98d24Cf4B2011842e64316Ff6d042c",
            _g[ChainId.POLYGON] = "0x565098CBa693b3325f9fe01D41b7A1cd792Abab1",
            _g[ChainId.FANTOM] = "0xa0554607e477cdC9d0EE2A6b087F4b2DC2815C22",
            _g[ChainId.BOBA] = "0x61A269a9506272D128d79ABfE8E8276570967f00",
            _g[ChainId.METIS] = "0x0b5740c6b4a97f90eF2F0220651Cca420B868FfB",
            _g[ChainId.MOONBEAM] = "0x5CF84397944B9554A278870B510e86667681ff8D",
            _g[ChainId.MOONRIVER] = "0xa9D0C0E124F53f4bE1439EBc35A9C73c0e8275fB",
            _g[ChainId.ARBITRUM] = "0x13780E6d5696DD91454F6d3BbC2616687fEa43d0",
            _g[ChainId.AVALANCHE] = "0xE97097dE8d6A17Be3c39d53AE63347706dCf8f43",
            _g[ChainId.DFK] = "0x360d6DD540E3448371876662FBE7F1aCaf08c5Ab",
            _g[ChainId.AURORA] = "0xb1Da21B0531257a7E5aEfa0cd3CbF23AfC674cE1",
            _g[ChainId.HARMONY] = "0xa0554607e477cdC9d0EE2A6b087F4b2DC2815C22",
            _g),
        swapType: SwapType.UST
    });
    // ETH, ETH wrappers, and nETH :D
    /**
     * ETH is the native currency ("gas token") for Ethereum, Optimism,
     * Boba, Arbitrum, and countless other chains
     * (though the four listed are the four currently supported by Synapse Protocol).
     * See {@link WETH} for a "wrapped" ERC20 variant of ETH.
     */
    Tokens.ETH = new BaseToken({
        name: "Ethereum",
        symbol: "ETH",
        decimals: 18,
        addresses: (_h = {},
            _h[ChainId.ETH] = "",
            _h[ChainId.OPTIMISM] = "",
            _h[ChainId.BOBA] = "",
            _h[ChainId.ARBITRUM] = "",
            _h),
        swapType: SwapType.ETH,
        isETH: true,
        isGasToken: true,
    });
    /**
     * nETH is a token involved in the bridge.
     */
    Tokens.NETH = new BaseToken({
        name: "Synapse nETH",
        symbol: "nETH",
        decimals: 18,
        addresses: (_j = {},
            _j[ChainId.OPTIMISM] = "0x809DC529f07651bD43A172e8dB6f4a7a0d771036",
            _j[ChainId.CRONOS] = "0x41E95B1F1c7849c50Bb9Caf92AB33302c0de945F",
            _j[ChainId.FANTOM] = "0x67C10C397dD0Ba417329543c1a40eb48AAa7cd00",
            _j[ChainId.BOBA] = "0x96419929d7949D6A801A6909c145C8EEf6A40431",
            _j[ChainId.MOONBEAM] = "0x3192Ae73315c3634Ffa217f71CF6CBc30FeE349A",
            _j[ChainId.METIS] = "0x931B8f17764362A3325D30681009f0eDd6211231",
            _j[ChainId.ARBITRUM] = "0x3ea9B0ab55F34Fb188824Ee288CeaEfC63cf908e",
            _j[ChainId.AVALANCHE] = "0x19E1ae0eE35c0404f835521146206595d37981ae",
            _j[ChainId.HARMONY] = "0x0b5740c6b4a97f90eF2F0220651Cca420B868FfB",
            _j),
        swapType: SwapType.ETH
    });
    Tokens.WETH = new BaseToken({
        name: "Wrapped ETH",
        symbol: "WETH",
        decimals: 18,
        addresses: (_k = {},
            _k[ChainId.ETH] = "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            _k[ChainId.OPTIMISM] = "0x121ab82b49B2BC4c7901CA46B8277962b4350204",
            _k[ChainId.BOBA] = "0xd203De32170130082896b4111eDF825a4774c18E",
            _k[ChainId.ARBITRUM] = "0x82af49447d8a07e3bd95bd0d56f35241523fbab1",
            _k),
        swapType: SwapType.ETH
    });
    Tokens.WETHBEAM = new BaseToken({
        name: "Moonbeam Wrapped ETH",
        symbol: "WETH",
        decimals: 18,
        addresses: (_l = {},
            _l[ChainId.MOONBEAM] = "0x3192Ae73315c3634Ffa217f71CF6CBc30FeE349A",
            _l),
        swapType: SwapType.ETH,
    });
    Tokens.WETH_E = new BaseToken({
        name: "Wrapped Ether",
        symbol: "WETH.e",
        decimals: 18,
        addresses: (_m = {},
            _m[ChainId.AVALANCHE] = "0x49d5c2bdffac6ce2bfdb6640f4f80f226bc10bab",
            _m),
        swapType: SwapType.ETH,
    });
    Tokens.AVWETH = new BaseToken({
        name: "AAVE Wrapped Ether",
        symbol: "AVWETH",
        decimals: 18,
        addresses: (_o = {},
            _o[ChainId.AVALANCHE] = "0x53f7c5869a859f0aec3d334ee8b4cf01e3492f21",
            _o),
        swapType: SwapType.ETH,
    });
    Tokens.ONE_ETH = new BaseToken({
        name: "Harmony ETH",
        symbol: "1ETH",
        decimals: 18,
        addresses: (_p = {},
            _p[ChainId.HARMONY] = "0x6983d1e6def3690c4d616b13597a09e6193ea013",
            _p),
        swapType: SwapType.ETH,
    });
    Tokens.FTM_ETH = new BaseToken({
        name: "Wrapped ETH",
        symbol: "ETH ",
        decimals: 18,
        addresses: (_q = {},
            _q[ChainId.FANTOM] = "0x74b23882a30290451A17c44f4F05243b6b58C76d",
            _q),
        swapType: SwapType.ETH,
    });
    Tokens.METIS_ETH = new BaseToken({
        name: "Wrapped ETH",
        symbol: "ETH",
        decimals: 18,
        addresses: (_r = {},
            _r[ChainId.METIS] = "0x420000000000000000000000000000000000000A",
            _r),
        swapType: SwapType.ETH,
    });
    // Synapse tokens
    Tokens.SYN = new BaseToken({
        name: "Synapse",
        symbol: "SYN",
        decimals: 18,
        addresses: (_s = {},
            _s[ChainId.ETH] = "0x0f2d719407fdbeff09d87557abb7232601fd9f29",
            _s[ChainId.OPTIMISM] = "0x5A5fFf6F753d7C11A56A52FE47a177a87e431655",
            _s[ChainId.CRONOS] = "0xFD0F80899983b8D46152aa1717D76cba71a31616",
            _s[ChainId.BSC] = "0xa4080f1778e69467e905b8d6f72f6e441f9e9484",
            _s[ChainId.POLYGON] = "0xf8f9efc0db77d8881500bb06ff5d6abc3070e695",
            _s[ChainId.FANTOM] = "0xE55e19Fb4F2D85af758950957714292DAC1e25B2",
            _s[ChainId.BOBA] = "0xb554A55358fF0382Fb21F0a478C3546d1106Be8c",
            _s[ChainId.METIS] = "0x67C10C397dD0Ba417329543c1a40eb48AAa7cd00",
            _s[ChainId.MOONBEAM] = "0xF44938b0125A6662f9536281aD2CD6c499F22004",
            _s[ChainId.MOONRIVER] = "0xd80d8688b02B3FD3afb81cDb124F188BB5aD0445",
            _s[ChainId.ARBITRUM] = "0x080f6aed32fc474dd5717105dba5ea57268f46eb",
            _s[ChainId.AVALANCHE] = "0x1f1E7c893855525b303f99bDF5c3c05Be09ca251",
            _s[ChainId.AURORA] = "0xd80d8688b02B3FD3afb81cDb124F188BB5aD0445",
            _s[ChainId.HARMONY] = "0xE55e19Fb4F2D85af758950957714292DAC1e25B2",
            _s),
        swapType: SwapType.SYN
    });
    /**
     * nUSD is a token involved in the bridge.
     */
    Tokens.NUSD = new BaseToken({
        name: "Synapse nUSD",
        symbol: "nUSD",
        decimals: 18,
        addresses: (_t = {},
            _t[ChainId.ETH] = "0x1B84765dE8B7566e4cEAF4D0fD3c5aF52D3DdE4F",
            _t[ChainId.OPTIMISM] = "0x67C10C397dD0Ba417329543c1a40eb48AAa7cd00",
            _t[ChainId.CRONOS] = "0x396c9c192dd323995346632581BEF92a31AC623b",
            _t[ChainId.BSC] = "0x23b891e5c62e0955ae2bd185990103928ab817b3",
            _t[ChainId.POLYGON] = "0xb6c473756050de474286bed418b77aeac39b02af",
            _t[ChainId.FANTOM] = "0xED2a7edd7413021d440b09D654f3b87712abAB66",
            _t[ChainId.BOBA] = "0x6B4712AE9797C199edd44F897cA09BC57628a1CF",
            _t[ChainId.METIS] = "0x961318Fc85475E125B99Cc9215f62679aE5200aB",
            _t[ChainId.ARBITRUM] = "0x2913E812Cf0dcCA30FB28E6Cac3d2DCFF4497688",
            _t[ChainId.AVALANCHE] = "0xCFc37A6AB183dd4aED08C204D1c2773c0b1BDf46",
            _t[ChainId.DFK] = "0x3AD9DFE640E1A9Cc1D9B0948620820D975c3803a",
            _t[ChainId.AURORA] = "0x07379565cD8B0CaE7c60Dc78e7f601b34AF2A21c",
            _t[ChainId.HARMONY] = "0xED2a7edd7413021d440b09D654f3b87712abAB66",
            _t),
        swapType: SwapType.USD,
    });
    Tokens.DFK_USDC = new BaseToken({
        name: "USD Circle",
        symbol: "USDC",
        decimals: 18,
        addresses: (_u = {},
            _u[ChainId.DFK] = "0x3AD9DFE640E1A9Cc1D9B0948620820D975c3803a",
            _u),
        swapType: SwapType.USD,
    });
    // chain native coins and wrapper tokens
    /**
     * AVAX is the native currency of Avalanche C-Chain.
     * See {@link WAVAX} for the "wrapped" ERC20 variant of this token.
     */
    Tokens.AVAX = new BaseToken({
        name: "Avalanche",
        symbol: "AVAX",
        decimals: 18,
        addresses: (_v = {},
            _v[ChainId.AVALANCHE] = "",
            _v),
        swapType: SwapType.AVAX,
        isGasToken: true,
    });
    Tokens.WAVAX = new WrapperToken({
        name: "Wrapped AVAX",
        symbol: "wAVAX",
        decimals: 18,
        addresses: (_w = {},
            _w[ChainId.MOONBEAM] = "0xA1f8890E39b4d8E33efe296D698fe42Fb5e59cC3",
            _w[ChainId.AVALANCHE] = "0xb31f66aa3c1e785363f0875a1b74e27b85fd66c7",
            _w[ChainId.DFK] = "0xB57B60DeBDB0b8172bb6316a9164bd3C695F133a",
            _w[ChainId.HARMONY] = "0xD9eAA386cCD65F30b77FF175F6b52115FE454fD6" // synAVAX, but here for compat.
        ,
            _w),
        swapType: SwapType.AVAX,
        underlyingToken: Tokens.AVAX,
    });
    /**
     * SYN_AVAX (synAVAX) is a "wrapped" ERC20 form of AVAX,
     * and is the supported output Token when bridging AVAX from
     * Avalanche C-Chain or Defi Kingdoms mainnet to Harmony.
     */
    Tokens.SYN_AVAX = new BaseToken({
        name: "Wrapped AVAX",
        symbol: "synAVAX",
        decimals: 18,
        addresses: (_x = {},
            _x[ChainId.HARMONY] = "0xD9eAA386cCD65F30b77FF175F6b52115FE454fD6",
            _x[ChainId.AVALANCHE] = "0xD9eAA386cCD65F30b77FF175F6b52115FE454fD6" // Not actually real.
        ,
            _x),
        swapType: SwapType.AVAX
    });
    Tokens.MULTI_AVAX = new BaseToken({
        name: "AnySwap/Multi Wrapped AVAX",
        symbol: "multiAVAX",
        decimals: 18,
        addresses: (_y = {},
            _y[ChainId.HARMONY] = "0xb12c13e66ade1f72f71834f2fc5082db8c091358",
            _y),
        swapType: SwapType.AVAX
    });
    /**
     * MOVR is the native currency of the Moonriver chain.
     * See {@link WMOVR} for the "wrapped" ERC20 variant of this token.
     */
    Tokens.MOVR = new BaseToken({
        name: "Moonriver",
        symbol: "MOVR",
        decimals: 18,
        addresses: (_z = {},
            _z[ChainId.MOONRIVER] = "",
            _z),
        swapType: SwapType.MOVR,
        isGasToken: true,
    });
    Tokens.WMOVR = new WrapperToken({
        name: "Wrapped MOVR",
        symbol: "wMOVR",
        decimals: 18,
        addresses: (_0 = {},
            _0[ChainId.MOONBEAM] = "0x1d4C2a246311bB9f827F4C768e277FF5787B7D7E",
            _0[ChainId.MOONRIVER] = "0x98878b06940ae243284ca214f92bb71a2b032b8a",
            _0),
        swapType: SwapType.MOVR,
        underlyingToken: Tokens.MOVR,
    });
    /**
     * GAS_JEWEL is the native currency of the DeFi Kingdoms chain.
     * See {@link JEWEL} for the "wrapped" ERC20 variant of this token.
     */
    Tokens.GAS_JEWEL = new BaseToken({
        name: "JEWEL",
        symbol: "JEWEL",
        decimals: 18,
        addresses: (_1 = {},
            _1[ChainId.DFK] = "",
            _1),
        swapType: SwapType.JEWEL,
        isGasToken: true,
    });
    /**
     * JEWEL is a "wrapped" ERC20 variant of {@link GAS_JEWEL}.
     * In the context of Synapse Protocol, JEWEL is primarily used for
     * bridging the native currency of the DeFi Kingdoms mainnet, {@link GAS_JEWEL},
     * to and from the Harmony and Avalanche networks.
     */
    Tokens.JEWEL = new WrapperToken({
        name: "JEWEL",
        symbol: "JEWEL",
        decimals: 18,
        addresses: (_2 = {},
            _2[ChainId.AVALANCHE] = "0x997Ddaa07d716995DE90577C123Db411584E5E46",
            _2[ChainId.DFK] = "0xCCb93dABD71c8Dad03Fc4CE5559dC3D89F67a260",
            _2[ChainId.HARMONY] = "0x72Cb10C6bfA5624dD07Ef608027E366bd690048F",
            _2),
        swapType: SwapType.JEWEL,
        underlyingToken: Tokens.GAS_JEWEL,
    });
    /**
     * SYN_JEWEL ("synJewel") is an ERC20 Token used by Synapse Protocol to
     * bridge and "wrap" the native currency of the DeFi Kingdoms mainnet, {@link GAS_JEWEL},
     * to and from Avalanche C-Chain.
     *
     * synJEWEL is supported on Harmony as an "intermediate" token; for example, it is
     * possible to bridge {@link JEWEL} from Harmony to synJEWEL on Avalanche, and then
     * bridge synJEWEL to native {@link GAS_JEWEL} on the DeFi Kingdoms mainnet.
     */
    Tokens.SYN_JEWEL = new BaseToken({
        name: "synJEWEL",
        symbol: "synJEWEL",
        decimals: 18,
        addresses: (_3 = {},
            _3[ChainId.HARMONY] = "0x28b42698Caf46B4B012CF38b6C75867E0762186D",
            _3),
        swapType: SwapType.JEWEL
    });
    Tokens.MULTIJEWEL = new BaseToken({
        name: "Multichain JEWEL",
        symbol: "multiJEWEL",
        decimals: 18,
        addresses: (_4 = {},
            _4[ChainId.AVALANCHE] = "0x4f60a160D8C2DDdaAfe16FCC57566dB84D674BD6",
            _4),
        swapType: SwapType.JEWEL
    });
    Tokens.XJEWEL = new BaseToken({
        name: "xJEWEL",
        symbol: "xJEWEL",
        decimals: 18,
        addresses: (_5 = {},
            _5[ChainId.DFK] = "0x77f2656d04E158f915bC22f07B779D94c1DC47Ff",
            _5[ChainId.HARMONY] = "0xA9cE83507D872C5e1273E745aBcfDa849DAA654F",
            _5),
        swapType: SwapType.XJEWEL
    });
    // non-Synapse, non-stablecoin tokens
    Tokens.GOHM = new BaseToken({
        name: "Olympus DAO",
        symbol: "gOHM",
        decimals: 18,
        addresses: (_6 = {},
            _6[ChainId.ETH] = "0x0ab87046fBb341D058F17CBC4c1133F25a20a52f",
            _6[ChainId.OPTIMISM] = "0x0b5740c6b4a97f90eF2F0220651Cca420B868FfB",
            _6[ChainId.CRONOS] = "0xbB0A63A6CA2071c6C4bcAC11a1A317b20E3E999C",
            _6[ChainId.BSC] = "0x88918495892BAF4536611E38E75D771Dc6Ec0863",
            _6[ChainId.POLYGON] = "0xd8cA34fd379d9ca3C6Ee3b3905678320F5b45195",
            _6[ChainId.FANTOM] = "0x91fa20244Fb509e8289CA630E5db3E9166233FDc",
            _6[ChainId.BOBA] = "0xd22C0a4Af486C7FA08e282E9eB5f30F9AaA62C95",
            _6[ChainId.METIS] = "0xFB21B70922B9f6e3C6274BcD6CB1aa8A0fe20B80",
            _6[ChainId.MOONBEAM] = "0xD2666441443DAa61492FFe0F37717578714a4521",
            _6[ChainId.MOONRIVER] = "0x3bF21Ce864e58731B6f28D68d5928BcBEb0Ad172",
            _6[ChainId.ARBITRUM] = "0x8D9bA570D6cb60C7e3e0F31343Efe75AB8E65FB1",
            _6[ChainId.AVALANCHE] = "0x321E7092a180BB43555132ec53AaA65a5bF84251",
            _6[ChainId.HARMONY] = "0x67C10C397dD0Ba417329543c1a40eb48AAa7cd00",
            _6),
        swapType: SwapType.OHM,
    });
    Tokens.HIGH = new BaseToken({
        name: "Highstreet",
        symbol: "HIGH",
        decimals: 18,
        addresses: (_7 = {},
            _7[ChainId.ETH] = "0x71Ab77b7dbB4fa7e017BC15090b2163221420282",
            _7[ChainId.BSC] = "0x5f4bde007dc06b867f86ebfe4802e34a1ffeed63",
            _7),
        swapType: SwapType.HIGH
    });
    Tokens.JUMP = new BaseToken({
        name: "HyperJump",
        symbol: "JUMP",
        decimals: 18,
        addresses: (_8 = {},
            _8[ChainId.BSC] = "0x130025ee738a66e691e6a7a62381cb33c6d9ae83",
            _8[ChainId.FANTOM] = "0x78DE9326792ce1d6eCA0c978753c6953Cdeedd73",
            _8),
        swapType: SwapType.JUMP
    });
    Tokens.DOG = new BaseToken({
        name: "The Doge NFT",
        symbol: "DOG",
        decimals: 18,
        addresses: (_9 = {},
            _9[ChainId.ETH] = "0xBAac2B4491727D78D2b78815144570b9f2Fe8899",
            _9[ChainId.BSC] = "0xaa88c603d142c371ea0eac8756123c5805edee03",
            _9[ChainId.POLYGON] = "0xeee3371b89fc43ea970e908536fcddd975135d8a",
            _9),
        swapType: SwapType.DOG
    });
    Tokens.NFD = new BaseToken({
        name: "Feisty Doge",
        symbol: "NFD",
        decimals: 18,
        addresses: (_10 = {},
            _10[ChainId.BSC] = "0x0fe9778c005a5a6115cbe12b0568a2d50b765a51",
            _10[ChainId.POLYGON] = "0x0a5926027d407222f8fe20f24cb16e103f617046",
            _10[ChainId.AVALANCHE] = "0xf1293574ee43950e7a8c9f1005ff097a9a713959",
            _10),
        swapType: SwapType.NFD,
    });
    // FRAX/synFrax
    Tokens.FRAX = new BaseToken({
        name: "Frax",
        symbol: "FRAX",
        decimals: 18,
        addresses: (_11 = {},
            _11[ChainId.ETH] = "0x853d955acef822db058eb8505911ed77f175b99e",
            _11[ChainId.MOONBEAM] = "",
            _11[ChainId.MOONRIVER] = "0x1a93b23281cc1cde4c4741353f3064709a16197d",
            _11[ChainId.HARMONY] = "0xFa7191D292d5633f702B0bd7E3E3BcCC0e633200",
            _11),
        swapType: SwapType.FRAX,
    });
    Tokens.SYN_FRAX = new BaseToken({
        name: "Synapse Frax",
        symbol: "synFRAX",
        decimals: 18,
        addresses: (_12 = {},
            _12[ChainId.FANTOM] = "0x1852F70512298d56e9c8FDd905e02581E04ddb2a",
            _12[ChainId.MOONBEAM] = "0xDd47A348AB60c61Ad6B60cA8C31ea5e00eBfAB4F",
            _12[ChainId.MOONRIVER] = "0xE96AC70907ffF3Efee79f502C985A7A21Bce407d",
            _12[ChainId.HARMONY] = "0x1852F70512298d56e9c8FDd905e02581E04ddb2a",
            _12),
        swapType: SwapType.FRAX,
    });
    Tokens.SOLAR = new BaseToken({
        name: "Vested SolarBeam",
        symbol: "veSOLAR",
        decimals: 18,
        addresses: (_13 = {},
            _13[ChainId.MOONBEAM] = "0x0DB6729C03C85B0708166cA92801BcB5CAc781fC",
            _13[ChainId.MOONRIVER] = "0x76906411D07815491A5E577022757aD941fb5066",
            _13),
        swapType: SwapType.SOLAR,
    });
    Tokens.GMX = new BaseToken({
        name: "GMX",
        symbol: "GMX",
        decimals: 18,
        addresses: (_14 = {},
            _14[ChainId.ARBITRUM] = "0xfc5a1a6eb076a2c7ad06ed22c90d7e710e35ad0a",
            _14[ChainId.AVALANCHE] = "0x62edc0692bd897d2295872a9ffcac5425011c661",
            _14),
        wrapperAddresses: (_15 = {},
            _15[ChainId.AVALANCHE] = "0x20A9DC684B4d0407EF8C9A302BEAaA18ee15F656",
            _15),
        swapType: SwapType.GMX,
    });
    Tokens.NEWO = new BaseToken({
        name: "New Order",
        symbol: "NEWO",
        decimals: 18,
        addresses: (_16 = {},
            _16[ChainId.ETH] = "0x98585dFc8d9e7D48F0b1aE47ce33332CF4237D96",
            _16[ChainId.ARBITRUM] = "0x0877154a755B24D499B8e2bD7ecD54d3c92BA433",
            _16[ChainId.AVALANCHE] = "0x4Bfc90322dD638F81F034517359BD447f8E0235a",
            _16),
        swapType: SwapType.NEWO,
    });
    Tokens.SDT = new BaseToken({
        name: "Stake DAO Token",
        symbol: "SDT",
        decimals: 18,
        addresses: (_17 = {},
            _17[ChainId.ETH] = "0x73968b9a57c6e53d41345fd57a6e6ae27d6cdb2f",
            _17[ChainId.FANTOM] = "0xE3c82A836Ec85311a433fBd9486EfAF4b1AFbF48",
            _17[ChainId.AVALANCHE] = "0xCCBf7c451F81752F7d2237F2c18C371E6e089E69",
            _17[ChainId.HARMONY] = "0xE3c82A836Ec85311a433fBd9486EfAF4b1AFbF48",
            _17),
        swapType: SwapType.SDT,
    });
    Tokens.LUNA = new BaseToken({
        name: "LUNA",
        symbol: "LUNA",
        decimals: 18,
        addresses: (_18 = {},
            _18[ChainId.OPTIMISM] = "0x931B8f17764362A3325D30681009f0eDd6211231",
            _18[ChainId.ARBITRUM] = "0x1A4dA80967373fd929961e976b4b53ceeC063a15",
            _18),
        swapType: SwapType.LUNA,
    });
    Tokens.USDB = new BaseToken({
        name: "USD Balance",
        symbol: "USDB",
        decimals: 18,
        addresses: (_19 = {},
            _19[ChainId.ETH] = "0x02b5453d92b730f29a86a0d5ef6e930c4cf8860b",
            _19[ChainId.BSC] = "0xc8699abbba90c7479dedccef19ef78969a2fc608",
            _19[ChainId.POLYGON] = "0xfa1fbb8ef55a4855e5688c0ee13ac3f202486286",
            _19[ChainId.FANTOM] = "0x6fc9383486c163fa48becdec79d6058f984f62ca",
            _19[ChainId.MOONRIVER] = "0x3e193c39626bafb41ebe8bdd11ec7cca9b3ec0b2",
            _19[ChainId.AVALANCHE] = "0x5ab7084cb9d270c2cb052dd30dbecbca42f8620c",
            _19),
        swapType: SwapType.USDB
    });
    Tokens.VSTA = new BaseToken({
        name: "Vesta",
        symbol: "VSTA",
        decimals: 18,
        addresses: (_20 = {},
            _20[ChainId.ETH] = "0xA8d7F5e7C78ed0Fa097Cc5Ec66C1DC3104c9bbeb",
            _20[ChainId.ARBITRUM] = "0xa684cd057951541187f288294a1e1c2646aa2d24",
            _20),
        swapType: SwapType.VSTA
    });
    Tokens.mintBurnTokens = [
        Tokens.NUSD, Tokens.SYN, Tokens.NETH,
        Tokens.HIGH, Tokens.DOG, Tokens.JUMP,
        Tokens.FRAX, Tokens.SYN_FRAX, Tokens.NFD,
        Tokens.GOHM, Tokens.SOLAR, Tokens.GMX,
        Tokens.UST, Tokens.NEWO, Tokens.SDT,
        Tokens.LUNA, Tokens.USDB, Tokens.SYN_AVAX,
        Tokens.GAS_JEWEL, Tokens.JEWEL, Tokens.SYN_JEWEL,
        Tokens.XJEWEL, Tokens.MULTIJEWEL, Tokens.DFK_USDC,
        Tokens.VSTA
    ];
    Tokens.isMintBurnToken = function (token) { return Tokens.mintBurnTokens.map(function (t) { return t.id; }).includes(token.id); };
    Tokens.ChainGasTokensMap = (_21 = {},
        _21[ChainId.ETH] = Tokens.ETH,
        _21[ChainId.OPTIMISM] = Tokens.ETH,
        _21[ChainId.BOBA] = Tokens.ETH,
        _21[ChainId.MOONRIVER] = Tokens.MOVR,
        _21[ChainId.ARBITRUM] = Tokens.ETH,
        _21[ChainId.AVALANCHE] = Tokens.AVAX,
        _21[ChainId.DFK] = Tokens.GAS_JEWEL,
        _21);
    /**
     * Returns the native currency Token ("gas token") for a given chain, if
     * it's supported by Synapse Protocol.
     * @param chainId
     */
    Tokens.gasTokenForChain = function (chainId) {
        if (chainId in Tokens.ChainGasTokensMap) {
            return Tokens.ChainGasTokensMap[chainId];
        }
        return null;
    };
    /**
     * Returns the "wrapper" Token for a given native currency Token ("gas token")
     * if such Token exists and is supported by Synapse Protocol.
     * @param gasToken
     */
    Tokens.gasTokenWrapper = function (gasToken) {
        if (!gasToken) {
            return null;
        }
        switch (tokenSwitch(gasToken)) {
            case Tokens.AVAX:
                return Tokens.WAVAX;
            case Tokens.MOVR:
                return Tokens.WMOVR;
            case Tokens.ETH:
                return Tokens.WETH;
            case Tokens.GAS_JEWEL:
                return Tokens.JEWEL;
        }
        return null;
    };
    /**
     * checkTokenAllowance returns the amount of `args.token` belonging to `args.owner`
     * which `args.spender` is allowed to spend on behalf of `args.owner`.
     * @param {CheckTokenAllowanceParams} args {@link CheckTokenAllowanceParams} object containing arguments
     * @param {number} args.chainId Chain ID of the network on which to fetch the spend allowance of `spender` for `owner`'s `token`.
     * @param {Token} args.token Token to fetch allowance information of.
     * @param {string} args.owner Address for owner of `token`
     * @param {string} args.spender Address for spender of `owner`'s `token`
     */
    function checkTokenAllowance(args) {
        return __awaiter(this, void 0, void 0, function () {
            var tokenParams;
            return __generator(this, function (_a) {
                tokenParams = {
                    chainId: args.chainId,
                    tokenAddress: args.token.address(args.chainId)
                };
                return [2 /*return*/, ERC20.allowanceOf(args.owner, args.spender, tokenParams)];
            });
        });
    }
    Tokens.checkTokenAllowance = checkTokenAllowance;
    /**
     * approveTokenSpend approves `args.spender` to spend `args.amount` (or the ERC20 max approval amount)
     * of `args.token` belonging to `args.signer`
     * @param {ApproveTokenParams} args {@link ApproveTokenParams} object containing arguments
     * @param {number} args.chainId Chain ID of the network on which to approve the spend allowance of `spender` for `signer`'s `token`
     * @param {Token} args.token Token to approve
     * @param {string} args.spender Address of spender to approve use of `signer`'s `token`
     * @param {BigNumber} args.amount [Optional] amount of `owner`'s `token` to approve for spend by `spender`. Defaults to uint256 max (0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) if not provided.
     *
     * @return {Promise<ContractTransaction>} Executed transaction object.
     */
    function approveTokenSpend(args) {
        return __awaiter(this, void 0, void 0, function () {
            var tokenParams, approveArgs;
            return __generator(this, function (_a) {
                tokenParams = {
                    tokenAddress: args.token.address(args.chainId),
                    chainId: args.chainId
                };
                approveArgs = {
                    spender: args.spender,
                    amount: args.amount
                };
                return [2 /*return*/, ERC20.approve(approveArgs, tokenParams, args.signer)];
            });
        });
    }
    Tokens.approveTokenSpend = approveTokenSpend;
    Tokens.AllTokens = [
        Tokens.DAI, Tokens.BUSD, Tokens.USDC, Tokens.USDT, Tokens.UST,
        Tokens.ETH, Tokens.NETH, Tokens.WETH, Tokens.WETHBEAM, Tokens.WETH_E,
        Tokens.AVWETH, Tokens.ONE_ETH, Tokens.FTM_ETH, Tokens.METIS_ETH,
        Tokens.SYN, Tokens.NUSD, Tokens.AVAX, Tokens.WAVAX, Tokens.MOVR, Tokens.WMOVR,
        Tokens.GOHM, Tokens.HIGH, Tokens.JUMP, Tokens.DOG, Tokens.NFD, Tokens.FRAX,
        Tokens.SYN_FRAX, Tokens.SOLAR, Tokens.GMX, Tokens.NEWO, Tokens.SDT,
        Tokens.LUNA, Tokens.USDB, Tokens.SYN_AVAX, Tokens.GAS_JEWEL, Tokens.JEWEL,
        Tokens.SYN_JEWEL, Tokens.XJEWEL, Tokens.MULTIJEWEL, Tokens.DFK_USDC,
        Tokens.MULTI_AVAX, Tokens.VSTA,
    ];
    /**
     * Returns a {@link Token} object based on the passed `tokenSymbol`, if such token exists.
     *
     * @param tokenSymbol Actual token symbol ("DAI", "SYN") or the `id` field of a {@link Token} object.
     *
     * @return A {@link Token} if one matches the passed `tokenSymbol`, null otherwise.
     */
    function tokenFromSymbol(tokenSymbol) {
        var res = null;
        for (var _i = 0, AllTokens_1 = Tokens.AllTokens; _i < AllTokens_1.length; _i++) {
            var t = AllTokens_1[_i];
            if (typeof tokenSymbol === 'symbol') {
                if (t.id === tokenSymbol) {
                    res = t;
                    break;
                }
            }
            else if (typeof tokenSymbol === 'string') {
                if (t.symbol === tokenSymbol) {
                    res = t;
                    break;
                }
            }
        }
        return res;
    }
    Tokens.tokenFromSymbol = tokenFromSymbol;
})(Tokens || (Tokens = {}));
