// @synapseprotocol/sdk/dist/internal/minirpc.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { fetchJson } from "@ethersproject/web";
var RequestError = /** @class */ (function (_super) {
    __extends(RequestError, _super);
    function RequestError(message, code, data) {
        var _this = _super.call(this, message) || this;
        _this.code = code;
        _this.data = data;
        _this.name = _this.constructor.name; // dafuq
        _this.message = message; // dafuq message
        return _this;
    }
    return RequestError;
}(Error));
export { RequestError };
var MiniRpcProvider = /** @class */ (function () {
    function MiniRpcProvider(chainId, url, batchWaitTimeMs) {
        if (batchWaitTimeMs === void 0) { batchWaitTimeMs = 50; }
        this.isMetaMask = false;
        this._nextId = 1;
        this._batchAggregator = null;
        this._pendingBatch = null;
        this.chainId = chainId;
        var parsed = new URL(url);
        this._url = parsed.toString();
        this._host = parsed.host;
        this._path = parsed.pathname;
        // how long to wait to batch calls
        this._batchInterval = batchWaitTimeMs;
    }
    Object.defineProperty(MiniRpcProvider.prototype, "batchInterval", {
        /* c8 ignore start */
        /**
         * Amount of time, in milliseconds, between batch RPC calls
         */
        get: function () {
            return this._batchInterval;
        },
        /**
         * Sets the provider's interval for sending batch RPC calls.
         * @param interval amount of time in milliseconds the provider will wait between sending batch RPC calls
         * @internal
         */
        set: function (interval) {
            this._batchInterval = interval;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MiniRpcProvider.prototype, "url", {
        get: function () {
            return this._url;
        },
        /**
         * @internal
         */
        set: function (newUrl) {
            var parsed = newUrl instanceof URL ? newUrl : new URL(newUrl);
            this._host = parsed.host;
            this._path = parsed.pathname;
            this._url = parsed.toString();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MiniRpcProvider.prototype, "host", {
        get: function () {
            return this._host;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MiniRpcProvider.prototype, "path", {
        get: function () {
            return this._path;
        },
        enumerable: false,
        configurable: true
    });
    /* c8 ignore stop */
    MiniRpcProvider.prototype.request = function (request) {
        return __awaiter(this, void 0, void 0, function () {
            var batchItem, prom;
            var _this = this;
            return __generator(this, function (_a) {
                if (request.method === 'eth_chainId') {
                    return [2 /*return*/, "0x".concat(this.chainId.toString(16))];
                }
                if (this._pendingBatch === null) {
                    this._pendingBatch = [];
                }
                batchItem = {
                    request: __assign({ jsonrpc: "2.0", id: (this._nextId++) }, request),
                    resolve: null,
                    reject: null,
                };
                prom = new Promise(function (resolve, reject) {
                    batchItem.resolve = resolve;
                    batchItem.reject = reject;
                });
                this._pendingBatch.push(batchItem);
                if (!this._batchAggregator) {
                    setTimeout(function () { return _this._processBatch(); }, this._batchInterval);
                }
                return [2 /*return*/, prom];
            });
        });
    };
    MiniRpcProvider.prototype._processBatch = function () {
        return __awaiter(this, void 0, void 0, function () {
            var currentBatch, requests;
            return __generator(this, function (_a) {
                currentBatch = this._pendingBatch;
                this._pendingBatch = null;
                this._batchAggregator = null;
                if (currentBatch === null) {
                    currentBatch = [];
                }
                requests = currentBatch.map(function (req) { return req.request; });
                if (requests.length === 0) {
                    return [2 /*return*/];
                }
                return [2 /*return*/, fetchJson(this._url, JSON.stringify(requests))
                        .then(function (result) {
                        return currentBatch.forEach(function (req, idx) {
                            var payload = result[idx];
                            if (payload) {
                                if (payload.error) {
                                    var _a = payload.error, message = _a.message, code = _a.code, data = _a.data;
                                    req.reject(new RequestError(message, code, data));
                                }
                                else {
                                    req.resolve(payload.result);
                                }
                            }
                        });
                    })
                        .catch(function (error) { return currentBatch.forEach(function (batchItem) { return batchItem.reject(error); }); })];
            });
        });
    };
    return MiniRpcProvider;
}());
export { MiniRpcProvider };
