// @synapseprotocol/sdk/dist/internal/rpcconnector.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { Web3Provider } from "@ethersproject/providers";
import { MiniRpcProvider } from "./minirpc.js";
import { AbstractConnector } from "@web3-react/abstract-connector";
var RpcConnector = /** @class */ (function () {
    function RpcConnector(args) {
        var _this = this;
        var urls = args.urls, _a = args.batchInterval, batchInterval = _a === void 0 ? 50 : _a;
        this._chainEndpoints = urls;
        var miniRpcProviders = Object.keys(urls).reduce(function (acc, chainId) {
            var cid = Number(chainId);
            acc[cid] = _this._newProvider(cid, urls[cid], batchInterval);
            return acc;
        }, {});
        var web3Providers = Object.keys(miniRpcProviders).reduce(function (acc, chainId) {
            var cid = Number(chainId);
            acc[cid] = new Web3Provider(miniRpcProviders[cid]);
            return acc;
        }, {});
        this._providers = miniRpcProviders;
        this._web3Providers = web3Providers;
    }
    RpcConnector.prototype.provider = function (chainId) {
        return this._web3Providers[chainId];
    };
    /**
     * @internal
     */
    RpcConnector.prototype.setProviderConfig = function (chainId, endpoint, batchInterval) {
        if (batchInterval === void 0) { batchInterval = 50; }
        var provider = this._newProvider(chainId, endpoint, batchInterval);
        delete this._chainEndpoints[chainId];
        delete this._providers[chainId];
        delete this._web3Providers[chainId];
        this._chainEndpoints[chainId] = endpoint;
        this._providers[chainId] = provider;
        this._web3Providers[chainId] = new Web3Provider(provider);
    };
    RpcConnector.prototype._newProvider = function (chainId, endpoint, batchInterval) {
        return new MiniRpcProvider(chainId, endpoint, batchInterval);
    };
    return RpcConnector;
}());
export { RpcConnector };
var NetworkConnector = /** @class */ (function (_super) {
    __extends(NetworkConnector, _super);
    function NetworkConnector(args) {
        var _this = this;
        var defaultChainId = args.defaultChainId, urls = args.urls;
        var supportedChainIds = Object.keys(urls).map(function (k) { return Number(k); });
        _this = _super.call(this, { supportedChainIds: supportedChainIds }) || this;
        _this.currentChainId = defaultChainId || Number(Object.keys(urls)[0]);
        _this.providers = Object.keys(urls).reduce(function (acc, chainId) {
            var cid = Number(chainId);
            acc[cid] = new MiniRpcProvider(cid, urls[cid]);
            return acc;
        }, {});
        return _this;
    }
    Object.defineProperty(NetworkConnector.prototype, "provider", {
        get: function () {
            return this.providers[this.currentChainId];
        },
        enumerable: false,
        configurable: true
    });
    NetworkConnector.prototype.activate = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, {
                        provider: this.providers[this.currentChainId],
                        chainId: this.currentChainId,
                        account: null
                    }];
            });
        });
    };
    NetworkConnector.prototype.getProvider = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.providers[this.currentChainId]];
            });
        });
    };
    NetworkConnector.prototype.getChainId = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.currentChainId];
            });
        });
    };
    NetworkConnector.prototype.getAccount = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, null];
            });
        });
    };
    NetworkConnector.prototype.deactivate = function () {
        return;
    };
    return NetworkConnector;
}(AbstractConnector));
export { NetworkConnector };
